### PR TITLE
Touch driver refactoring

### LINKED ISSUE #3507: Touch panel events may be lost
There are few problems with implementation and processing touch panel event on T-model. 

1) When the user touches the display, touch controller starts generating events with 11ms period. The firmware
should readout those events from the touch controller registers via I2C interface. If the firmware doesn't readout the event, the controller overwrites it withby the succesive one => If the firmware is busy and is not calling touch_read() frequently enough, 
events may be lost and it's especially true for the `TOUCH_START` event.

2) The current `touch_read()` implementation discards all events if succesive calls doesn't occur within less than 50ms in order to not provide outdated events to UI components. This behaviour might lead in loosing events as well.

We didn't observe this behaviour until I began experimenting with animations on the display (when call to `touch_read()` are less frequent). 

We need to find out more suitable solution to both A) avoid loosing touch events and B) prevent providing outdated events to the UI.

Suggestions:
A) Process touch panel events within the interrupt routine. 
B) Add a timestamp to the touch event and process events only if their creating time is later than the time the target component became visible to the user.

### LINKED ISSUE #3732: Separate touch and button implementation in trezorhal/unix.
These are mangled together now but should be separated as stm hal.

### PR DESCRIPTION
**Pull request by cepetr**
This PR focuses on refactoring the touch screen controller drivers. It does not introduce new features but nearly fixes #3507 issue (so we can probably close it). 

1. Removed `lib/touch.c` since it's unused.
2. Polished the touch driver API (`touch.h`), resulting in fewer functions with clearer purposes. Also added the `touch_deinit()` function, allowing repeated driver initialization and deinitialization.
3. Refactored all drivers (emulator, discovery kits and FT6X36) to avoid using a many global variables, resulting in the driver state being clear and well-defined in a single location.
4. Heavily refactored the FT6X36 driver - removing duplicated code and simplifying complicated logic without affecting behavior.
5. Split the touch driver on the simulator into `touch.c` and `button.c` modules. (fixed #3732 )
6. Fixed a bug introduced in #3932 - `TOUCH VERSION` is functional again

The original goal was to prepare the touch controller driver for power management features (e.g., suspending before entering low power modes and resuming after returning to full power).

### CODE REVIEW

File: core/embed/trezorhal/stm32f4/touch/stmpe811.c
Comment (TychoVrahe) [inline core/embed/trezorhal/stm32f4/touch/stmpe811.c:9]: typo

File: core/embed/trezorhal/stm32f4/touch/stmpe811.c
Comment (TychoVrahe) [inline core/embed/trezorhal/stm32f4/touch/stmpe811.c:569]: static

File: core/embed/trezorhal/stm32u5/touch/sitronix.c
Comment (TychoVrahe) [inline core/embed/trezorhal/stm32u5/touch/sitronix.c:1147]: static

File: core/embed/trezorhal/touch.h
Comment (TychoVrahe) [inline core/embed/trezorhal/touch.h:41]: currently

File: core/embed/trezorhal/unix/button.c
Comment (TychoVrahe) [inline core/embed/trezorhal/unix/button.c:22]: could we remove this ifdef in favor of only compiling this file for 'board' that support buttons? (likewise for USE_TOUCH)

File: core/embed/trezorhal/unix/touch.c
Comment (TychoVrahe) [inline core/embed/trezorhal/unix/touch.c:76]: driver can be pointer to const

File: core/embed/trezorhal/unix/touch.c
Comment (TychoVrahe) [inline core/embed/trezorhal/unix/touch.c:68]: static

File: core/embed/trezorhal/unix/touch.c
Comment (TychoVrahe) [inline core/embed/trezorhal/unix/touch.c:226]: lets keep calling it emulator in code/comments

File: core/embed/trezorhal/unix/touch.c
Comment (TychoVrahe) [inline core/embed/trezorhal/unix/touch.c:231]: lets keep calling it emulator in code/comments

**Review by TychoVrahe — COMMENTED** [review]
Some comments below. 

Didn't look too deep into discovery kit changes as they are not very important. Let me know if you think i should review it thoroughly

File: core/embed/trezorhal/stm32f4/touch/stmpe811.c
Comment (cepetr) [reply core/embed/trezorhal/stm32f4/touch/stmpe811.c]: fixed, see c04b50822941a7c6f979e8923827122438fe5749

File: core/embed/trezorhal/unix/touch.c
Comment (cepetr) [reply core/embed/trezorhal/unix/touch.c]: fixed, see c04b50822941a7c6f979e8923827122438fe5749

File: core/embed/trezorhal/unix/touch.c
Comment (cepetr) [reply core/embed/trezorhal/unix/touch.c]: fixed, see c04b50822941a7c6f979e8923827122438fe5749

File: core/embed/trezorhal/unix/touch.c
Comment (cepetr) [reply core/embed/trezorhal/unix/touch.c]: yes, fixed, see 38441e0bc9e4a44a5a2cafff09a1118ca5142d7d

File: core/embed/trezorhal/stm32f4/touch/ft6x36.c
Comment (cepetr) [reply core/embed/trezorhal/stm32f4/touch/ft6x36.c]: yes, it was weird idea. I removed it, see d641553c28edb27ab6bb424b34db6c01aa061d2f

File: core/embed/trezorhal/stm32f4/touch/stmpe811.c
Comment (cepetr) [reply core/embed/trezorhal/stm32f4/touch/stmpe811.c]: fixed, see a25e0d9312fd8a3398dcffab1f649eed50441c5f

File: core/embed/trezorhal/stm32u5/touch/sitronix.c
Comment (cepetr) [reply core/embed/trezorhal/stm32u5/touch/sitronix.c]: fixed, see a25e0d9312fd8a3398dcffab1f649eed50441c5f

File: core/embed/trezorhal/unix/touch.c
Comment (cepetr) [reply core/embed/trezorhal/unix/touch.c]: fixed, see a25e0d9312fd8a3398dcffab1f649eed50441c5f

File: core/embed/trezorhal/unix/button.c
Comment (cepetr) [reply core/embed/trezorhal/unix/button.c]: I've just removed these ifdefs, see 31b0bf98500a63a40288b8e7a441be82edd27394. Any changes in the build script are  not necessary.

File: core/embed/trezorhal/stm32f4/touch/ft6x36.c
Comment (cepetr) [reply core/embed/trezorhal/stm32f4/touch/ft6x36.c:408]: I also considered this and have to admit I am not absolutely sure. On one side, this branch blocks sending `TOUCH_END` twice (it should not happen because of the interrupt line check, but who knows...). On the other side, it is really missed the `PRESS_DOWN` event. But if it is, then the touch duration was really short—just 10ms (with no `CONTACT` event in between).

I would rather keep it as it is since the proper solution requires simulating `TOUCH_START` and subsequent `TOUCH_END` in the next call, which is more complicated and probably brings no real value. We can of course discuss it further.

File: core/embed/bootloader/main.c
Comment (cepetr) [reply core/embed/bootloader/main.c:532]: In my opinion, the previous timeout might be too short to be on the safe side. According to the specs, we have to wait at least 300ms until the touch controller is ready. If somebody is touching the screen, it starts reporting events with a ~10ms period. Currently, we wait 50ms, so we have a chance to detect up to 5 events. Before, there was just a 10ms wait. I understand it worked before as well; maybe the 300ms requirement has a large margin. We can shorten it, but I am not sure that it's a good idea.

File: core/embed/trezorhal/stm32f4/touch/ft6x36.c
Comment (TychoVrahe) [reply core/embed/trezorhal/stm32f4/touch/ft6x36.c:408]: I am fine with that, just please adjust the comment so its clear that its intentional

File: core/embed/trezorhal/unix/button.c
Comment (TychoVrahe) [reply core/embed/trezorhal/unix/button.c]: Not necessary for sure, but should lower the compilation time both locally and also in CI, where we build a lot (though it remains very questionable whether it has some real effect), and its more consistent with how it is done in HW builds

File: core/embed/bootloader/main.c
Comment (TychoVrahe) [reply core/embed/bootloader/main.c:532]: yeah i know of zero issues with the previous setting. Otoh 40ms difference is negligible so we can stay on the safe side no prob.

File: core/embed/trezorhal/stm32f4/touch/ft6x36.c
Comment (cepetr) [reply core/embed/trezorhal/stm32f4/touch/ft6x36.c]: commented, see a96ade8a4b2b371145185d554c3ce3ff0feb8299

File: core/embed/trezorhal/unix/button.c
Comment (cepetr) [reply core/embed/trezorhal/unix/button.c]: I agree with this, but I believe such a change requires some preparation. Currently, `USE_TOUCH` and `USE_BUTTON` are defined in header files, so we cannot use them in build scripts. This situation suggests we should open a discussion about the consistent use of feature flags across C, Rust, and build scripts.

Comment (TychoVrahe) [reply core/embed/trezorhal/unix/button.c]: The build scripts are actually quite ready for this, see https://github.com/trezor/trezor-firmware/commit/7a8e2535c0ab930a8ea7bfcd7ff8744392f0fd5f - feel free to cherrypick. 

Yes we could also define `USE_TOUCH` and `USE_BUTTON` there or discuss some other solutions, but its not needed for this particular change.

Comment (cepetr) [reply core/embed/trezorhal/unix/button.c]: Nice, I didn't know about this. Cherry picked: 50f67d53db1934e2df49ddcc923e87ccbd96419e

**Review by TychoVrahe — APPROVED** [review]
lgtm, there are some style issues so please make sure to fix them before merging

**Pull request merged**
